feature/add national page tests

oftheheadland opened the pull request:
## Main Changes:
* Add tests for the National page

## Steps To Test:
1. `npm run cypress` in client directory

cschwinderg commented on app/client/cypress/integration/national/national.spec.js:
```diff
@@ -0,0 +1,188 @@
+      .contains(
+        'of our rivers and streams are healthy based on biological communities',
+      )
+      .should('not.exist');
```
Similar to above.

cschwinderg commented on app/client/cypress/integration/national/national.spec.js:
```diff
@@ -0,0 +1,188 @@
+    cy.get('h2')
+      .filter(':visible')
+      .contains('Explore National Drinking Water Information')
+      .should('exist');
```
For this I'd suggest doing something similar to what discussed during the testing meeting and put a "data-testid" on each panel. Then you can do 

```suggestion
    cy.findByTestId('<the-id>')
       .findByText('Explore National Drinking Water Information')
       .should('be.visible');
```

cschwinderg commented on app/client/cypress/integration/national/national.spec.js:
```diff
@@ -0,0 +1,188 @@
+    cy.get('button')
+      .filter(':visible')
+      .contains('Expand All')
+      .should('exist');
```
Similar to above.

cschwinderg commented on app/client/cypress/integration/national/national.spec.js:
```diff
@@ -0,0 +1,188 @@
+    cy.findByText('National Drinking Water').click();
+    cy.get('#tabs--2--panel--1 > div > table')
+      .should('exist')
+      .and('be.visible');
```
Similar to above.

oftheheadland commented:
@cschwinderg thanks, I'll look into these items

oftheheadland commented on app/client/cypress/integration/national/national.spec.js:
```diff
@@ -0,0 +1,189 @@
+    cy.findAllByTestId('hmw-expand-collapse')
+      .filter(':visible')
+      .contains('Expand All')
+      .should('exist');
```
@cschwinderg for the other requested changes I adjusted the pattern of getting elements and filtering them to instead use data-testids.  

But for these Expand/Collapse All buttons I ran into an issue, There are 4 of these buttons on the National page because they come from the Accordion component. I'm using FindAllByTestId and then filtering by visible items, which ends up being the same number of lines. I think this is still an improvement but I'm open to any ideas you have

cschwinderg replied:
This looks good to me, since it is no longer getting all of the buttons on the screen. 

Thank you for making the changes.

courtneymyers commented on app/client/cypress/integration/national/national.spec.js:
```diff
@@ -0,0 +1,189 @@
+    cy.findByText('EPA has posted information through this', {
+      exact: false,
+    }).should('not.exist');
+  });
```
@oftheheadland I know we're chatting on MS Teams, but you should take a look at a similar test Caleb wrote:

https://github.com/Eastern-Research-Group/mywaterway/blob/develop/app/client/cypress/integration/state/links.spec.js#L22

Probably best to be consistent if we can on:
1) prefer regex over string w/ `exact: false`
2) store regex/strings you're searching for in a variable

oftheheadland commented:
@courtneymyers I've applied the changes from our meeting, let me know what you think

courtneymyers commented on app/client/cypress/integration/national.spec.js:
```diff
@@ -0,0 +1,188 @@
+      .filter(':visible')
+      .contains('Expand All')
+      .should('exist');
+    cy.findByText(text).should('not.exist');
```
You could simplify this test quite a bit. In general only assert what you need to assert (in this case, that the accordion content is visible or not). A unit test would be better suited to make sure the accordion button text changes every time, but it's not difficult to roll that small assertion in as you're doing...just no need to assert both that "Collapse All" is shown and "Expand All" is not. Here's my suggestion on simplifying this, while still testing what you're wanting to test:

```suggestion
    const text = /^Biological condition tells us how healthy a waterbody is/;

    // verify that accordion is initially collapsed
    cy.findAllByText('Expand All').filter(':visible');
    cy.findByText(text).should('not.exist');

    // click button, and verify that accordion is expanded
    cy.findAllByText('Expand All')
      .filter(':visible')
      .click()
      .contains('Collapse All');
    cy.findByText(text).should('be.visible');

    // click button again, and verify that accordion is collapsed
    cy.findByText('Collapse All').click();
    cy.findByText(text).should('not.exist');
```

courtneymyers commented on app/client/cypress/integration/national.spec.js:
```diff
@@ -0,0 +1,188 @@
+      .should('have.attr', 'target', '_blank');
+    cy.get('a[href*="https://www.epa.gov/nutrientpollution"]')
+      .filter(':visible')
+      .should('have.attr', 'rel', 'noopener noreferrer');
```
Here's where regex helps you. There aren't multiple cases on the page where "Nutrient Pollution" exists (with that capitalization), other than in the link. As always, prefer `dom-testing-library` selectors over cypress `get()` (notice a theme here 😉).

```suggestion
    cy.findByText(/Nutrient Pollution/)
      .should('have.attr', 'href', 'https://www.epa.gov/nutrientpollution')
      .should('have.attr', 'target', '_blank')
      .should('have.attr', 'rel', 'noopener noreferrer');
```

courtneymyers commented on app/client/cypress/integration/national.spec.js:
```diff
@@ -0,0 +1,188 @@
+
+    cy.findByText(
+      /^Learn more about what EPA is doing to reduce nutrient pollution/,
+    ).should('have.attr', 'rel', 'noopener noreferrer');
```
I realized `.should()` returns its subject, so you can chain subsequent `.should()` calls to simplify your code:

```suggestion
    cy.findByText(
      /Learn more about what EPA is doing to reduce nutrient pollution/,
    )
      .should(
        'have.attr',
        'href',
        'https://www.epa.gov/nutrient-policy-data/what-epa-doing-reduce-nutrient-pollution',
      )
      .should('have.attr', 'target', '_blank')
      .should('have.attr', 'rel', 'noopener noreferrer');
```

courtneymyers commented on app/client/cypress/integration/national.spec.js:
```diff
@@ -0,0 +1,188 @@
+    ).should('have.attr', 'target', '_blank');
+    cy.get(
+      'a[href*="https://www.epa.gov/national-aquatic-resource-surveys/national-rivers-and-streams-assessment-2008-2009-results"]',
+    ).should('have.attr', 'rel', 'noopener noreferrer');
```
This one's a bit different... The user's really clicking an image (with a caption text) – they don't have any idea what the URL is, other than hovering over the linked image. The ideal way you'd select images  is with dom-testing-library's `findByAltText` query – that has the added benefit of making sure you're using alt text for images in the first place.

Here's a situation where I think it's ok to traverse the DOM w/ Cypress's `parent()` method a few times to get to the surrounding anchor. Normally you wouldn't want to do that, as it goes against the whole philosophy of the user not knowing or caring what the DOM structure looks like (and makes refactoring the structure of your markup break tests, which is not ideal). But in this case, it's probably the closest way to replicate what a user would be doing (looking at at an image, and clicking it). It could be argued that it'd be even better targeting the caption as well, but I'm good w/ this simpler approach:

```suggestion
    cy.findByAltText(/Learn more about Rivers & Streams/)
      .parent()
      .parent()
      .should(
        'have.attr',
        'href',
        'https://www.epa.gov/national-aquatic-resource-surveys/national-rivers-and-streams-assessment-2008-2009-results',
      )
      .should('have.attr', 'target', '_blank')
      .should('have.attr', 'rel', 'noopener noreferrer');
```

courtneymyers commented on app/client/cypress/integration/national.spec.js:
```diff
@@ -0,0 +1,188 @@
+      'noopener noreferrer',
+    );
+  });
+});
```
This test belongs in the other `describe` block (other tab). It also should follow the same things I mentioned above: don't use `cy.get()` and chain `should()`s.

Here's a simplified version of the block, but it still needs to be moved to the other `describe` block:
```suggestion
  it('Clicking the “Consumer Confidence Report (CCR)” link opens a new tab.', () => {
    // since Cypress doesn't support multiple tabs, we'll do the next best thing
    // (https://docs.cypress.io/guides/references/trade-offs.html#Multiple-tabs)
    cy.findByText('Consumer Confidence Report (CCR)')
      .should('have.attr', 'href', 'https://www.epa.gov/ccr')
      .should('have.attr', 'target', '_blank')
      .should('have.attr', 'rel', 'noopener noreferrer');
  });
});
```

courtneymyers commented on app/client/src/components/shared/Accordion/index.js:
```diff
@@ -108,7 +108,10 @@ function AccordionList({
         {!expandDisabled && (
-          <ExpandButton onClick={(ev) => setAllExpanded(!allExpanded)}>
+          <ExpandButton
+            data-testid="hmw-expand-collapse"
```
This `data-testid` attribute is no longer needed.

oftheheadland commented:
@courtneymyers thank you for the review, I've cleaned up the tests, moved a couple around, fixed the sub-tab test that wasn't accomplishing its goal, and removed most of the .get()s

courtneymyers commented:
> @courtneymyers thank you for the review, I've cleaned up the tests, moved a couple around, fixed the sub-tab test that wasn't accomplishing its goal, and removed most of the .get()s

Thanks, @oftheheadland. Can you update the "Clicking a Glossary Term..." test to not use `cy.get()` as well?
